google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/CredentialsService.java

```diff
@@ -0,0 +1,163 @@
+   * @return the {@link GoogleCredentials} object pointed to by the URL.
+   * @throws SpannerException If the URL does not point to a valid credentials file, or if the file
+   *     cannot be accessed.
+   */
+  GoogleCredentials createCredentials(String credentialsUrl) {
+    try {
+      if (credentialsUrl == null) {
+        return internalGetApplicationDefault();
+      } else {
+        return getCredentialsFromUrl(credentialsUrl);
+      }
+    } catch (IOException e) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Invalid credentials path specified", e);
+    }
+  }
+
+  @VisibleForTesting
+  GoogleCredentials internalGetApplicationDefault() throws IOException {
+    return GoogleCredentials.getApplicationDefault();
+  }
+
+  private GoogleCredentials getCredentialsFromUrl(String credentialsUrl) throws IOException {
+    Preconditions.checkNotNull(credentialsUrl);
+    Preconditions.checkArgument(
+        credentialsUrl.length() > 0, "credentialsUrl may not be an empty string");
+    if (credentialsUrl.startsWith(GOOGLE_CLOUD_STORAGE_PREFIX)) {
+      return getCredentialsFromCloudStorage(credentialsUrl);
+    } else {
+      return getCredentialsFromLocalFile(credentialsUrl);
+    }
+  }
+
+  private GoogleCredentials getCredentialsFromLocalFile(String filePath) throws IOException {
+    File credentialsFile = new File(filePath);
+    if (!credentialsFile.isFile()) {
+      throw new IOException(
+          String.format("Error reading credential file %s: File does not exist", filePath));
+    }
+    try (InputStream credentialsStream = new FileInputStream(credentialsFile)) {
+      return GoogleCredentials.fromStream(
+          credentialsStream, CloudSpannerOAuthUtil.HTTP_TRANSPORT_FACTORY);
+    }
+  }
+
+  private GoogleCredentials getCredentialsFromCloudStorage(String credentialsUrl)
+      throws IOException {
+    Preconditions.checkArgument(credentialsUrl.startsWith(GOOGLE_CLOUD_STORAGE_PREFIX));
+    try {
+      Storage storage = internalCreateStorage();
+      String bucketName = internalGetBucket(credentialsUrl);
+      String blobName = internalGetBlob(credentialsUrl);
+      Blob blob = storage.get(bucketName, blobName);
+      return GoogleCredentials.fromStream(
+          internalCreateInputStream(blob), CloudSpannerOAuthUtil.HTTP_TRANSPORT_FACTORY);
+    } catch (Exception e) {
+      throw new IOException(
+          String.format(
+              "Could not retrieve credentials from %s. "
+                  + "Credentials on Google Cloud Storage can only be used when the application is running "
+                  + "in an environment that already has default credentials to access Google Cloud Storage",
+              credentialsUrl),
+          e);
+    }
+  }
+
+  @VisibleForTesting
+  Storage internalCreateStorage() {
+    return StorageOptions.newBuilder().build().getService();
```

Comment:
If we're defining our own `HttpTransport`/`HttpTransportFactory` then we should probably be using them here as well. 

But, since we're not actually defining anything custom (at least for now), we could probably use the defaults across the board (i.e. `GoogleCredentials.fromStream(InputStream)` instead of `GoogleCredentials.fromStream(InputStream, HttpTransportFactory)` and drop the `CloudSpannerOAuthUtil` inner class.

Reply:
There was a reason why it was made that way, but I cannot remember why (it's 9 months ago). I've just tested without the custom HttpTransport and it seems to work, so I've removed it.

Reply:
Eventually, it could be possible that a user needs a custom HttpTransport (proxy or other rules).